### DISCUSSION TITLE
Routed events and view models

### DISCUSSION THREAD

**Discussion by FraserElectronics** [AvaloniaUI/Avalonia#15441] [Q&A]
I have created a simple UserControl which has routed event and a button. When the button is clicked, I raise the routed event.

The event:

```
public static readonly RoutedEvent<RoutedEventArgs> MenuExpandingEvent = RoutedEvent.Register<MainMenuControl, RoutedEventArgs>(nameof(MenuExpanding), RoutingStrategies.Direct );
    public event EventHandler<RoutedEventArgs> MenuExpanding
    {
        add => AddHandler(MenuExpandingEvent, value);
        remove => RemoveHandler(MenuExpandingEvent, value);
    }
```

Raising the event:

```
RoutedEventArgs args = new RoutedEventArgs( MenuExpandingEvent, mb);
RaiseEvent( args );
```

When I try and bind a handler in my view model to this event, I get the error:

```
Unable to find suitable setter or adder for property MenuExpanding of type MyProject:MyProject.Components.MainMenuControl for argument Avalonia.Markup.Xaml:Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindingExtension
```

My AXAML looks like:

```
<components:MainMenuControl MenuExpanding="{Binding OnMenuExpanding}"/>
```

And my view model has the method:

```
public void OnMenuExpanding( object sender, RoutedEventArgs e )
{

}
```

Is there something I am doing wrong here?

Andy

**Comment by Pratham95VP** [comment]
I think routed events are not directly supported in viewmodels in Avalonia. Instead, you can handle the routed event in the code-behind of your view and then invoke a method in your view model.

**Comment by FraserElectronics** [comment]
@Pratham95VP That was going to be my fallback solution but I still get the same error even when the handler is in the code behind !

I now have:

```
<components:MainMenuControl MenuExpanding="OnMenuExpanding"/>
```

and in my code behind:

```
    public void OnMenuExpanding( object sender, RoutedEventArgs e )
    {
        if ( this.DataContext is IMainViewModel vm )
        {
            vm.OnMenuExpanding( sender, e );
        }
    }
```

but I get the error:

```
Error		Unable to find suitable setter or adder for property MenuExpanding of type MyProject:MyProject.Components.MainMenuControl for argument System.Private.CoreLib:System.String, available setter parameter lists are:
System.EventHandler`1[[Avalonia.Interactivity.RoutedEventArgs, Avalonia.Base, Version=11.0.2.0, Culture=neutral, PublicKeyToken=c8d484a7012f9a8b]]
```

**Reply by workgroupengineering** [reply]
You should add these methods to your custom control

```c#
public MainMenuControl 
{
     public void AddMenuExpandingHandler(Interactive element, EventHandler<RoutedEventArgs> handler) =>
          element.AddHandler(MenuExpandingEvent, handler);

     public void RemoveMenuExpandingHandler(Interactive element, EventHandler<RoutedEventArgs> handler) =>
          element.RemoveHandler(MenuExpandingEvent, handler);

}
``` 

or try  PR #15274

**Reply by FraserElectronics** [reply]
@workgroupengineering Not really sure how I am supposed to use that code?!
I now have:

```
   public event EventHandler<RoutedEventArgs> MenuExpanding
   {
       add => AddMenuExpandingHandler(this,value);
       remove => RemoveMenuExpandingHandler(this,value);
   }
```

Not sure what element is supposed to be - perhaps *this* is not correct?
In any case, I still get the same error "Unable to find a suitable setter or adder..."

Andy

**Reply by workgroupengineering** [reply]
MenuExpanding should be the same as before

```c#
public event EventHandler<RoutedEventArgs> MenuExpanding
    {
        add => AddHandler(MenuExpandingEvent, value);
        remove => RemoveHandler(MenuExpandingEvent, value);
    }
```

**Reply by FraserElectronics** [reply]
Sorry, I don't know how that helps :-(
How, when do AddMenuExpandingHandler / RemoveMenuExpandingHandler get called?
In any case, I still get the same compile error.

Andy

**Reply by workgroupengineering** [reply]
Sorry I forgot the static modifier.
```c#
public MainMenuControl 
{
     public static void AddMenuExpandingHandler(Interactive element, EventHandler<RoutedEventArgs> handler) =>
          element.AddHandler(MenuExpandingEvent, handler);

     public static void RemoveMenuExpandingHandler(Interactive element, EventHandler<RoutedEventArgs> handler) =>
          element.RemoveHandler(MenuExpandingEvent, handler);

}
```
>How, when do AddMenuExpandingHandler / RemoveMenuExpandingHandler get called?
In any case, I still get the same compile error.

They are called automatically by the xaml compiler

**Comment by FraserElectronics** [comment]
Okay, Avalonia is really messed up :-(
I did a clean and rebuild and still the same error then closed down VS and restarted, rebuilt and now the code behind is working !!!
It is still not my preferred solution as it requires the view to know the type of its view model and not sure that breaks the IOC principal.
In any case, I can work with this for now.

BR

Andy

**Reply by maxkatz6** [reply]
If rebuilding project and restarting IDE helps, it's probably not a Avalonia, but yeah, sucks.

> It is still not my preferred solution as it requires the view to know the type of its view model and not sure that breaks the IOC principal.

In MVVM your View knows type of ViewModel either way. Either it's using Bindings or code-behind view logic - conceptually it's the same, but different forms. You might prefer to keep all View logic in XAML though - that's where libraries like https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors usually help. They are not necessary, but really help doing XAML-first views.

**Comment by FraserElectronics** [comment]
Well I think something in Avalonia is not getting rebuilt because I quite often find that if I build and get an error, a second build clears the error. 
You are right about the view knowing about the view model as in my code behind constructor, I use the service container to get the view model so it *does* need to know about the view model. I am happy now with the solution but would still be nicer if somehow you could bind the routed event direct to the view model if that is something that is do-able in the future?

Thanks.

Andy

**Reply by timunie** [reply]
#15373 maybe?

**Reply by FraserElectronics** [reply]
Interesting 🤔
I will keep an eye on this PR 👍

Andy
